### PR TITLE
Fb-1816 : CharIndex str func

### PR DESCRIPTION
**Pull request by HHans09**
SQL string function CharIndex to identify the position of the substring in the input string. Position starts at 0

### CODE REVIEW

File: sql3/planner/inbuiltfunctionsstring.go [hunk 1]
```diff
@@ -644,3 +668,65 @@ func (n *callPlanExpression) EvaluateReplicate(currentRow []interface{}) (interf
 	return strings.Repeat(stringArg, int(intArg)), nil
 
 }
+
+// function to get the index of substring in the given string
```
Comment (rachithrr) [inline sql3/planner/inbuiltfunctionsstring.go:672]: The exported functions should follow this format
``` EvalutateCharIndex ... [and then describe what it does and any information that would be useful for a user or another developer to know] ```
More info about that here: https://go.dev/blog/godoc

File: sql3/planner/inbuiltfunctionsstring.go [hunk 2]
```diff
@@ -644,3 +668,65 @@ func (n *callPlanExpression) EvaluateReplicate(currentRow []interface{}) (interf
...
+	argEval, err := n.args[0].Evaluate(currentRow)
+	if err != nil {
+		return nil, err
+	}
+	if argEval == nil {
+		return nil, nil
+	}
+	subString, ok := argEval.(string)
+	if !ok {
+		return nil, sql3.NewErrInternalf("unexpected type converion %T", argEval)
```
Comment (rachithrr) [inline sql3/planner/inbuiltfunctionsstring.go:684]: fb-1817 adds a new function that wraps around ErrInternalF. Rebase once that is merged.

File: sql3/planner/inbuiltfunctionsstring.go [hunk 3]
```diff
@@ -644,3 +668,65 @@ func (n *callPlanExpression) EvaluateReplicate(currentRow []interface{}) (interf
...
+		return nil, nil
+	}
+	inputString, ok := argEval.(string)
+	if !ok {
+		return nil, sql3.NewErrInternalf("unexpected type converion %T", argEval)
+	}
+
+	if len(n.args) == 2 {
+		res := strings.Index(inputString, subString)
+		if res > -1 {
```
Comment (rachithrr) [inline sql3/planner/inbuiltfunctionsstring.go:701]: do we need this if condition? 
// return int64(res), nil

File: sql3/test/defs/defs_string_functions.go [hunk 4]
```diff
@@ -820,5 +820,87 @@ var stringScalarFunctionsTests = TableTest{
...
+				"select charindex('is','this is great','you')",
+			),
+			ExpErr: "integer expression expected",
+		},
+		{
+			name: "CharIndexIncorrectTypeofArgumentsforInputString",
+			SQLs: sqls(
+				"select charindex('is',23,3)",
+			),
+			ExpErr: "string expression expected",
```
Comment (rachithrr) [inline sql3/test/defs/defs_string_functions.go:843]: add compareexactordered

File: sql3/planner/inbuiltfunctionsstring.go (hunk 1 above)
Comment (HHans09) [reply sql3/planner/inbuiltfunctionsstring.go:672]: Updated

File: sql3/planner/inbuiltfunctionsstring.go (hunk 3 above)
Comment (HHans09) [reply sql3/planner/inbuiltfunctionsstring.go:701]: yes, when position is not found then it has to return -1

File: sql3/test/defs/defs_string_functions.go (hunk 4 above)
Comment (HHans09) [reply sql3/test/defs/defs_string_functions.go:843]: Do we need to add compareexactordered when we are expecting errors?

Comment (travisturner) [reply sql3/test/defs/defs_string_functions.go:843]: no, you don't need it for errors

File: sql3/planner/inbuiltfunctionsstring.go (hunk 3 above)
Comment (rachithrr) [reply sql3/planner/inbuiltfunctionsstring.go:701]: The strings.Index() returns -1 if not found, which is what we are returning in line 705.

File: sql3/planner/inbuiltfunctionsstring.go (hunk 2 above)
Comment (HHans09) [reply sql3/planner/inbuiltfunctionsstring.go]: Rebased and updated

**Review by rachithrr — APPROVED** [review]
looks good to me

**Pull request merged**